lib/puppet/type/user.rb

```diff
@@ -493,25 +493,6 @@ def exists?
       provider.exists?
     end
 
-    def retrieve
-      absent = false
-      properties.inject({}) { |prophash, property|
-        current_value = :absent
-
-        if absent
-          prophash[property] = :absent
-        else
-          current_value = property.retrieve
-          prophash[property] = current_value
-        end
-
-        if property.name == :ensure and current_value == :absent
-          absent = true
-        end
-        prophash
-      }
-    end
-
```

Comment:
I'm generally 👍 on this, and acceptance tests passed on a variety of platforms. Might be good to double check our tests around `Puppet::Type::User#retrieve` vs `Puppet::Type#retrieve` to see how the behavior changes.

At first glance, the overridden user retrieve returns a `Hash` while service retrieve returns `Puppet::Resource`:

```
irb(main):025:0> Puppet::Type.type(:user).new(name: 'root', groups: 'wheel', comment: 'something').retrieve.class
=> Hash
irb(main):026:0> Puppet::Type.type(:service).new(name: 'puppet').retrieve.class
=> Puppet::Resource
```

The `Puppet::Type#retrieve` method is called by `Puppet::Type#retrieve_resource`, and that accepts either `Hash` or `Puppet::Resource`, converting the former into the latter: https://github.com/puppetlabs/puppet/blob/bc5de2cb067c839ff03ffb5bcf43a1a6b3fbba81/lib/puppet/type.rb#L1132-L1134

I think one question is whether the `Hash` returned by the old code is equivalent to the `Puppet::Resource` returned by the new code?

A second question is whether `Puppet::Type::User#retrieve` is API? If so, then we can't make this change until Puppet 7.

Reply:
Debugged the methods old vs new and the `Puppet::Type#retrieve_resource` returns the same `Puppet::Resource`. Conversion of the Hash looks OK,  also did some manual tests on the user type and everything working as before.